### DISCUSSION TITLE
Best way to automatically reconnect

### DISCUSSION THREAD

**Discussion by aplinefresh** [tomer8007/kik-bot-api-unofficial#211] [Q&A]
Hi
I'm trying to work out the best way to automatically reconnect if the connection to kik is lost. I'm struggling to work out how to do this safely - any pointers?

**Comment by StethoSaysHello** [comment]
Hey Alpinefresh!

You can attempt to trigger one of the callbacks in a timed loop to check if the bot is online, and have it try to log in again if its not online. Basically just "refreshing" every few minutes.
I've posted an example [here.](https://pastebin.com/4C5JSk5E)
I used `on_chat_message_received` in the example but you can use [other callbacks](https://github.com/tomer8007/kik-bot-api-unofficial/blob/new/kik_unofficial/callbacks.py) too, it doesn't matter.

If your bots are highly active (frequently receiving group/private messages), you can reduce downtime by listening for ALL callbacks and having it only trigger "refreshing" if none of them have been triggered within a shorter timespan, like 1-2 minutes.

Hope this helps!

**Reply by aplinefresh** [reply]
Hi Stetho!
Thanks for your help. However it seems like when it gets logged off or the internet disconnects the script ends?

**Comment by Sitiaro** [comment]
Make an automation script to restart your bot every 10-15 mins. It doesn't require an internet connection or anything so your bot getting disconnected won't be a problem. If you've frequent internet issues then adjust the time in which the bot restarts to like 5 mins, and your bot script will auto-execute every 5 mins.
Here's what I've been using:-

```import os
os.system('pip install ./kik-bot-api-unofficial')
os.system('pip install schedule')
from subprocess import *
import schedule
import time

def auto_run():
    os.system('python echo_bot.py')

schedule.every(5).minutes.do(auto_run)
while 1:
    schedule.run_pending()
    time.sleep(1)

**Reply by aplinefresh** [reply]
Thanks for this, unfortunately even that amount of downtime is too much for what Im doing. It can take about 30 seconds to start up, so even with it at 5 minutes I can end up with a large chunk of that down.

**Comment by Jumvah** [comment]
Within the main bot project class I use the on_disconnect(self) function to count disconnects.
First disconnect is during authentication, any second disconnect is an actual disconnection from the server.
Eg:

```
class My_Project_Bot(KikClientCallback):

    def on_disconnected(self):
        global Disconnects
        Disconnects += 1
        if Disconnects == 1:
            print("Reconnecting for Authentication...\n")
        else:
            print("Connection Failed...Restarting...\n")
        return
```


Then at the end of my code where main() and is called, I have this:

```
while __name__ == __main__:
    Now = datetime.now()
    Date_Time = Now.strftime("%Y/%m/%d %H:%M:%S")
    print("Starting " + botname +" - " + Date_Time + "\n" )
    main()
    time.sleep(5)
   
    while Disconnects < 2:
        time.sleep(.1)

   Disconnects = 0
```


This seems to work fine. Also I send the stanza 
`"</ping>"` between 10 and 60 seconds apart and that seems to keep connection alive better.
